### ISSUE TITLE
tf.case evaluating all outputs when using batches (?)

### ISSUE THREAD

**Issue by MicaelCarvalho** [tensorflow/tensorflow#8168]
Hello,

I may be missing something here, but I couldn't find any useful information on the documentation, so I'm posting this behavior as a bug, hoping someone can clarify what's going on. Questions at the end of this issue

**Operating System:** Debian 4.8.15-2
**Installed version of CUDA and cuDNN:** CUDA 8, cuDNN 5
**python3 -c "import tensorflow; print(tensorflow.__version__)"**: 1.0.0

### Reproducible example

```
import tensorflow as tf

a = ([tf.constant(i) for i in range(2)],[tf.constant(i) for i in range(2)])
b = ([tf.constant(i) for i in range(2)],[tf.constant(i) for i in range(2)])

q1 = tf.train.slice_input_producer(a, num_epochs=3, shuffle=True, capacity=4)
q2 = tf.train.batch(q1, batch_size=2, num_threads=1, enqueue_many=False, capacity=4, allow_smaller_final_batch=True)

q3 = tf.train.slice_input_producer(b, num_epochs=1, shuffle=True, capacity=4)
q4 = tf.train.batch(q3, batch_size=2, num_threads=1, enqueue_many=False, capacity=4, allow_smaller_final_batch=True)

q4p = [tf.Print(q4[0], ['q4a']), tf.Print(q4[1], ['q4b'])]
q2p = [tf.Print(q2[0], ['q2a']), tf.Print(q2[1], ['q2b'])]

def get_op1():
	print('call1')
	return tf.Print(q2p, ['op1'])

def get_op2():
	print('call2')
	return tf.Print(q4p, ['op2'])

switcher = tf.placeholder(tf.bool)
test = tf.case([(switcher, get_op1)], default=get_op2, exclusive=True)

init = [tf.global_variables_initializer(), tf.local_variables_initializer()]
sess = tf.Session()
coord = tf.train.Coordinator()
sess.run(init)
threads = tf.train.start_queue_runners(coord=coord, sess=sess)

ignore = sess.run(test, feed_dict={switcher: True})

wait = input("So far, so good. We're evaluating q1/q2, but now q3/q4 will run out of examples. Press enter to continue.")

ignore = sess.run(test, feed_dict={switcher: True})
```

### Output

```
call2
call2
call1
[None, None]
I tensorflow/core/kernels/logging_ops.cc:79] [q4b]
I tensorflow/core/kernels/logging_ops.cc:79] [q2a]
I tensorflow/core/kernels/logging_ops.cc:79] [q4a]
I tensorflow/core/kernels/logging_ops.cc:79] [q2b]
I tensorflow/core/kernels/logging_ops.cc:79] [op1]
(some warnings/errors)
So far, so good. We're evaluating q1/q2, but now q3/q4 will run out of examples. Press enter to continue. (enter pressed)
(big errors, queue is empty)
```

### Questions

1) Why is `get_op2()` called 2 times before init?
2) Why are there so many messages after the first `sess.run(test...)`? And why do they say `OutOfRange`, when I have enough examples in the queue? (I can only imagine it's because the threads feeding the queue are dying, but weird nevertheless)
3) Although `op2` is not printed, `q4a` and `q4b` are, and they shouldn't, since we're evaluating `q2`. This led me to believe examples are being pulled from `q4` even though I didn't ask for them, and that is why the second `sess.run(test...)` crashes : no more examples in `q4`.

It seems tf.case is evaluating everything (weirdly, without printing `op2`), because `q4` goes to exhaustion without being used at all. I am using this system to switch between train/validation sets, but I'm being restricted by the amount of images in validation, which doesn't make any sense. So far I can only imagine either there is a bizarre bug on `tf.case` or I misunderstood something. Could you clarify this please?

Thanks in advance.

**Comment by MicaelCarvalho** [comment]
Update: This issue is related to #3287, which was closed without a proper solution (IMHO). The documentation doesn't explain why it runs all branches, and the questions 1 and 2 in this topic aren't addressed.

**Comment by yaroslavvb** [comment]
Conditional execution needs all ops to be created inside the the op. You created `q2p = [tf.Print(q2[0]` ahead of time, so those ops are executed regardless of the condition.

There's some extra explanation by mrry here -- http://stackoverflow.com/questions/34401714/is-tensorflow-lazy

**Comment by MicaelCarvalho** [comment]
Thanks for the link, I didn't know that -- the documentation should explain this behavior, it's definitely very important.

> Conditional execution needs all ops to be created inside the the op.

Does this mean conditional execution is, by nature, incompatible with anything that involves queues, since we cannot simply redefine the queue each time? -- in this case, redefine the queue = calling slice_input_producer and batch function.

If so, I don't see a way to solve the problem of having a batch variable with examples coming from two different sources. The only way to do so would be by putting a placeholder as input for the model, and manually evaluating the batch before passing it in the feed_dict parameter, effectively dividing the graph and treating the condition _à la python_. In my opinion, not an elegant solution.

Is there any way (_or, better, a correct way_) around this specific problem of selecting the set of images to be used?

**Comment by yaroslavvb** [comment]
Yes, it seems doing lazy dequeue with tf.cond is problematic. There's `tf.train.maybe_batch` op that can take over some of the target use cases

**Comment by MicaelCarvalho** [comment]
I didn't know `tf.train.maybe_batch` neither, but I don't think it can solve the problem. A quick read on its doc shows:

> keep_input: A bool Tensor. This tensor controls whether the input is added to the queue or not. If it is a scalar and evaluates True, then tensors are all added to the queue. If it is a vector and enqueue_many is True, then each example is added to the queue only if the corresponding value in keep_input is True. This tensor essentially acts as a filtering mechanism.

It seems the `keep_input` parameter can only control `enqueue`, whereas I'm trying to control `dequeue` (i.e. do not `dequeue` from validation if we're training, and do not `dequeue` from train if we're validating).

In my understanding, part of the philosophy behind TensorFlow is to have a well built graph, without having to manually control these things — like calling dequeue from two different places based on a "python condition", but rather embed it all into the graph and let TF optimize its way through it. In this perspective, not being able to implement a source selector in the graph sounds like a major fault.

Since I don't see a possible solution for this, I'm going to switch to the manual conditioning. Thank you very much for the help @yaroslavvb, I learned a lot here. :-)

PS: Leaving the issue open for a member's response and/or as a feature suggestion.

**Comment by MicaelCarvalho** [comment]
Just to be more precise and maybe help people who are looking for a solution:

The proposal/request in this thread is for something similar to `tf.QueueBase.from_list`, but receiving `dequeue` operations and/or the output of `tf.train.batch` instead of Queue objects. If one has manual control over the Queues, it is possible to use this method to pick which queue we should dequeue from using a single input variable. Without access to the Queue objects (as when using `tf.train.batch`), this is not possible.

_Maybe_ it's possible to _hack_ our way through the issue by retrieving the Queue objects from the graph and calling `tf.QueueBase.from_list` with them, but this would not be a clean solution.

**Comment by inflation** [comment]
Using multiple queues always bothers me.

A better option is discussed here #7951.

**Comment by ed-alertedh** [comment]
@MicaelCarvalho I seem to have implemented the hack you described. Fairly dirty, but if anybody wants a stopgap...

```
...
batch_train=tf.train.shuffle_batch(...)
batch_test=tf.train.shuffle_batch(...)
queue_selector = tf.placeholder(tf.int32, name='queue_selector')
train_q = tf.QueueBase(queue_ref=batch_train.op.inputs[0], dtypes=[tf.string], shapes=None, names=None)
test_q = tf.QueueBase(queue_ref=batch_test.op.inputs[0], dtypes=[tf.string], shapes=None, names=None)
selected_batch = tf.QueueBase.from_list(queue_selector, [train_q, test_q])
features = tf.parse_example(selected_batch)
```

**Comment by YoungBim** [comment]
@MicaelCarvalho I'mworking on the exact same issue and wonder if from your minimal code example you could gently write the "manual" strategy that you described ... Thanks in advance

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by michaelisard** [comment]
Closing as obsolete since we are moving from queues to datasets, but please reopen if this needs attention.
